More embed block fixes

notnownikki opened the pull request:
* Loading the rendered html on loading a saved block moved into componentWillMount
* Removed '.includes' usage, as it's not polyfilled
* Added a new rendering state for when we're loading a saved block
* Replaced string manipulation of urls to extract the host, with url.parse

notnownikki committed:
More embed block fixes

* Loading the rendered html on loading a saved block moved into componentWillMount
* Removed '.includes' usage, as it's not polyfilled
* Added a new rendering state for when we're loading a saved block
* Replaced string manipulation of urls to extract the host, with url.parse

mtias commented on blocks/library/embed/index.js:
```diff
@@ -118,17 +144,24 @@ registerBlockType( 'core/embed', {
 
+			if ( loadingFromSavedBlock ) {
+				// we're loading from a saved block, but haven't fetched the HTML yet...
+				return (
```
@jasmussen what do you think should be the UX here? With this change it'd be the url as text shown until we can load the generated embed. Maybe a white box saying "Generating embed..." with the url in smaller text could be better?

mtias commented on blocks/library/embed/index.js:
```diff
@@ -118,17 +144,24 @@ registerBlockType( 'core/embed', {
 
 		render() {
-			const { html, type, error, fetching } = this.state;
+			const { html, type, error, fetching, loadingFromSavedBlock } = this.state;
```
I think ideally this should be the single `loading` state, and it doesn't matter whether you just input the url in the embed field, or if it's loading from saved content.

notnownikki replied:
I think we need @jasmussen to cast a design spell on the loading state so it's suitable for both purposes :smile:

jasmussen committed:
Add some loading polish.

notnownikki committed:
Single "fetching content" UI

Use the same UI state for fetching new embedded content and
loading saved embed blocks in a post

notnownikki committed:
Set the value of the url input so that errors can be corrected if embedding fails

aduth commented on blocks/library/embed/index.js:
```diff
@@ -85,8 +91,14 @@ registerBlockType( 'core/embed', {
 			this.noPreview = [
 				'facebook.com',
```
I couldn't find much background on why we need to disable previews for some hosts. Should this be performed on the server instead of on the client? Also doesn't seem to need to be an instance variable assigned to each occurrence of the component but rather a top-level shared constant reference e.g. `PREVIEW_HOST_BLACKLIST`.

notnownikki replied:
Eventually we'll move the embedded previews in the editor into sandbox iframes. Facebook embeds refused to run inside these iframes, and throw up errors, so we want to display a "sorry, preview not available" message instead.

Though, yeah, noPreview should be top level and named better. Will fix!

aduth commented on blocks/library/embed/index.js:
```diff
@@ -10,6 +10,11 @@ import './style.scss';
+/**
+ * External dependencies
+ */
+import { parse } from 'url';
```
Minor: There's a bit of an undocumented ([semi-documented](https://github.com/WordPress/gutenberg/blob/master/docs/coding-guidelines.md#imports)?) convention around ordering these blocks `External` -> `WordPress` -> `Internal` (from most external to least external).

aduth commented on blocks/library/embed/index.js:
```diff
@@ -76,6 +81,7 @@ registerBlockType( 'core/embed', {
 		constructor() {
 			super( ...arguments );
 			this.doServerSideRender = this.doServerSideRender.bind( this );
+			this.isPreviewBlacklisted = this.isPreviewBlacklisted.bind( this );
```
Technically this function may not need to be bound because it's never called in a context that would have a `this` binding anything other than the current component instance (usually callbacks). But I think it's probably wise to err on the side of safety and `bind` every component function.

notnownikki replied:
This is getting removed now that the noPreview domains are getting slightly reworked.

aduth commented on blocks/library/embed/index.js:
```diff
@@ -129,31 +154,38 @@ registerBlockType( 'core/embed', {
+				return (
+					<div className="blocks-embed__loading">
+						<Spinner />
+						<p className="blocks-embed__loading-text">{ wp.i18n.__( 'Embedding...' ) }</p>
```
Minor: Ellipsis character `…` is more semantic than three dots for this purpose.

aduth commented on blocks/library/embed/index.js:
```diff
@@ -129,31 +154,38 @@ registerBlockType( 'core/embed', {
 
-			const domain = url.split( '/' )[ 2 ].replace( /^www\./, '' );
-			const cannotPreview = this.noPreview.includes( domain );
+			const urlBits = parse( url );
```
I might recommend a more meaningful name like `parsedUrl` than "bits" here.

notnownikki replied:
But "bits" is cute! Sure, will change :)

notnownikki commented on blocks/library/embed/index.js:
```diff
@@ -96,6 +108,19 @@ registerBlockType( 'core/embed', {
+				return false;
+			}
+			host = host.replace( 'www.', '' );
+			for ( let i = 0; i < this.noPreview.length; i++ ) {
```
oEmbed allows the user to embed with and without the www, even though facebook puts the www back in the embed html when its scripts run. So we need to account for that here.

notnownikki committed:
More fixes...

* Elipsis character instead of ...
* Removed cuteness from parsed url variable name
* Moved "no preview" list of domains into top level const
* Used lodash include
* Fixed the www. removing regex
* Fixed order of imports

notnownikki committed:
Removed redundant noPreview includes method

aduth commented on blocks/library/embed/index.js:
```diff
@@ -129,31 +140,38 @@ registerBlockType( 'core/embed', {
 
+			if ( fetching ) {
+				return (
+					<div className="blocks-embed__loading">
```
The docs could probably do for some clarification, but it's intended that the root element of the return value should simply be `top-level-folder` + `folder-name`,  i.e. `blocks-embed`. The treatment here is one reserved for descendants, or other components in the same folder (not `index.js`).

https://github.com/WordPress/gutenberg/blob/master/docs/coding-guidelines.md#naming

Would be more appropriate to apply loading as a modifier class:

```jsx
<div className="blocks-embed is-loading">
```

aduth reviewed (changes requested):
Good to go after modifier class change 👍

notnownikki committed:
Reorg classes for the embed fetching UI
